src/crimson/osd/main.cc

```diff
@@ -0,0 +1,101 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <iostream>
+
+#include <seastar/core/app-template.hh>
+#include <seastar/core/thread.hh>
+
+#include "common/ceph_argparse.h"
+#include "crimson/common/config_proxy.h"
+#include "crimson/net/SocketMessenger.h"
+
+#include "osd.h"
+
+using config_t = ceph::common::ConfigProxy;
+
+void usage(const char* prog) {
+  std::cout << "usage: " << prog << " -i <ID>" << std::endl;
+  generic_server_usage();
+}
+
+int main(int argc, char* argv[])
+{
+  std::vector<const char*> args{argv + 1, argv + argc};
+  if (ceph_argparse_need_usage(args)) {
+    usage(argv[0]);
+    return EXIT_SUCCESS;
+  }
+  std::string cluster;
+  std::string conf_file_list;
+  // ceph_argparse_early_args() could _exit(), while local_conf() won't ready
+  // until it's started. so do the boilerplate-settings parsing here.
+  auto init_params = ceph_argparse_early_args(args,
+                                              CEPH_ENTITY_TYPE_OSD,
+                                              &cluster,
+                                              &conf_file_list);
+  // TODO: add heartbeat
+  seastar::app_template app;
+  // talk with osd
+  ceph::net::SocketMessenger cluster_msgr("cluster", getpid());
+  // talk with mon/mgr
+  ceph::net::SocketMessenger client_msgr("client", getpid());
+  seastar::sharded<OSD> osd;
+
+  args.insert(begin(args), argv[0]);
+  try {
+    return app.run_deprecated(args.size(), const_cast<char**>(args.data()), [&] {
+        return seastar::async([&cluster_msgr, &client_msgr, &osd,
```

Comment:
why we run seastar::async here, it start a separate seastar::thread, here is our main work thread right?  for some timer work such as heartbeat we can use seastar::async to create a new seastar::thread to execute it.

Reply:
> why we run seastar::async here, it start a separate seastar::thread, here is our main work thread right?

i just like its simplicity. so we don't need to chain the continuations with `then()`. i don't follow you, could you define "main work thread"?

> for some timer work such as heartbeat we can use seastar::async to create a new seastar::thread to execute it.

periodical events can also triggered by `seastar::timer<>`, i think it's simpler.

Reply:
"A seastar thread blocking point is any function that returns a future<>. you block by calling future<>::get(); this waits for the future to become available, and in the meanwhile, other seastar threads and seastar non-threaded code may execute."  so when we call .get() if the future is not ready, the code will block there, is this our goal?

Reply:
> so when we call .get() if the future is not ready, the code will block there

yes. i think it's pretty much equivalent to chained continuation in this context. anyway, i changed it to chained `then()`, hope it could get rid of your confusion.